File: backend/integrations/project_ingest.py
"""Regras de ingest de repositório (espelho de frontend/js/project-ingest.js)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

PROJECT_MAP_NAME = "__project_map.txt"
MAX_CONTENT_FILES = 12
MAX_FILE_BYTES = 200 * 1024
MAX_CONTENT_CHARS = 200_000
MAX_MAP_LINES = 4000
MAX_ATTACHMENTS = MAX_CONTENT_FILES + 1

IGNORE_DIRS = frozenset(
    {
        "node_modules",
        ".git",
        "dist",
        "build",
        "vendor",
        "__pycache__",
        ".next",
        "coverage",
        ".venv",
        "venv",
        "target",
        "out",
        ".turbo",
        ".cache",
        ".idea",
        ".vscode",
        "eggs",
        ".eggs",
        ".tox",
        ".mypy_cache",
        ".ruff_cache",
        ".pytest_cache",
    }
)

IGNORE_EXT = frozenset(
    {
        ".png",
        ".jpg",
        ".jpeg",
        ".gif",
        ".webp",
        ".ico",
        ".svg",
        ".woff",
        ".woff2",
        ".ttf",
        ".eot",
        ".mp3",
        ".mp4",
        ".wav",
        ".zip",
        ".gz",
        ".tar",
        ".7z",
        ".rar",
        ".exe",
        ".dll",
        ".so",
        ".dylib",
        ".bin",
        ".pdf",
        ".pyc",
        ".class",
        ".o",
        ".a",
        ".wasm",
        ".map",
    }
)

HIGH_NAMES = frozenset(
    {
        "package.json",
        "package-lock.json",
        "pnpm-lock.yaml",
        "yarn.lock",
        "requirements.txt",
        "requirements-lock.txt",
        "pyproject.toml",
        "poetry.lock",
        "cargo.toml",
        "go.mod",
        "go.sum",
        "pom.xml",
        "composer.json",
        "composer.lock",
        "gemfile",
        "dockerfile",
        "docker-compose.yml",
        "docker-compose.yaml",
        "vercel.json",
        "next.config.js",
        "next.config.mjs",
        "next.config.ts",
        "nuxt.config.ts",
        "nuxt.config.js",
        "vite.config.ts",
        "vite.config.js",
        "tsconfig.json",
        "nginx.conf",
        ".env.example",
        ".env.sample",
        "openapi.yaml",
        "openapi.json",
        "swagger.yaml",
        "swagger.json",
    }
)

SOURCE_EXT = frozenset(
    {
        ".py",
        ".js",
        ".ts",
        ".tsx",
        ".jsx",
        ".go",
        ".rs",
        ".java",
        ".kt",
        ".rb",
        ".php",
        ".cs",
        ".swift",
        ".c",
        ".cpp",
        ".h",
        ".hpp",
        ".vue",
        ".svelte",
        ".sql",
        ".sh",
        ".ps1",
        ".yml",
        ".yaml",
        ".toml",
        ".ini",
        ".cfg",
        ".conf",
        ".json",
        ".md",
        ".html",
        ".css",
        ".scss",
    }
)


def normalize_path(path: str) -> str:
    text = str(path or "").replace("\\", "/").lstrip("./").lstrip("/")
    return text


def path_ignored(path: str) -> bool:
    norm = normalize_path(path)
    parts = [p for p in norm.split("/") if p]
    if not parts:
        return True
    for p in parts[:-1]:
        if p in IGNORE_DIRS:
            return True
    base = parts[-1]
    lower = base.lower()
    dot = lower.rfind(".")
    if dot >= 0 and lower[dot:] in IGNORE_EXT:
        return True
    if lower.endswith(".min.js") or lower.endswith(".min.css"):
        return True
    return False


def score_path(path: str, size: int = 0) -> int:
    norm = normalize_path(path)
    if not norm or path_ignored(norm):
        return -1
    if size > MAX_FILE_BYTES:
        return -1

    parts = norm.split("/")
    base = (parts[-1] or "").lower()
    score = 10

    if base in HIGH_NAMES:
        score += 100
    if base.startswith("dockerfile"):
        score += 90
    if base.startswith("docker-compose"):
        score += 90
    if base.endswith(".env.example") or base.endswith(".env.sample"):
        score += 85
    if "nginx" in base:
        score += 70

    segs = set(parts)
    if segs & {"routes", "api", "auth", "middleware", "controllers", "handlers"}:
        score += 60
    if segs & {"src", "app", "backend", "frontend", "server", "web"}:
        score += 25
    if len(parts) <= 2:
        score += 20

    dot = base.rfind(".")
    ext = base[dot:] if dot >= 0 else ""
    if ext in SOURCE_EXT:
        score += 15

    if segs & {"test", "tests", "__tests__", "spec", "docs", "doc", "examples", "fixtures"}:
        score -= 40
    if base.endswith(".test.js") or base.endswith(".spec.ts") or base.endswith("_test.py"):
        score -= 35
    if base.endswith(".lock") or base in {"yarn.lock", "pnpm-lock.yaml"}:
        score -= 50

    return score


@dataclass
class MapResult:
    text: str
    kept_count: int
    ignored_count: int
    total_seen: int


def build_project_map(
    entries: list[dict[str, Any]], *, max_lines: int = MAX_MAP_LINES
) -> MapResult:
    kept: list[tuple[str, int]] = []
    ignored = 0
    for e in entries:
        path = normalize_path(str(e.get("path") or ""))
        if not path or path_ignored(path):
            ignored += 1
            continue
        kept.append((path, int(e.get("size") or 0)))
    kept.sort(key=lambda x: x[0])
    shown = kept[:max_lines]
    omitted = len(kept) - len(shown)
    lines = [
        f"# Mapa do repositório ({len(kept)} arquivos após filtros; {ignored} ignorados)",
        "# Amostra de conteúdo anexada à parte (arquivos prioritários para pentest).",
        "",
    ]
    for path, size in shown:
        lines.append(f"{path}\t{size}")
    if omitted > 0:
        lines.append(f"… (+{omitted} omitidos do mapa)")
    return MapResult(
        text="\n".join(lines),
        kept_count=len(kept),
        ignored_count=ignored,
        total_seen=len(entries),
    )


def pick_content_paths(
    entries: list[dict[str, Any]], *, limit: int = MAX_CONTENT_FILES
) -> list[dict[str, Any]]:
    scored: list[tuple[int, str, int]] = []
    for e in entries:
        path = normalize_path(str(e.get("path") or ""))
        size = int(e.get("size") or 0)
        sc = score_path(path, size)
        if sc < 0:
            continue
        scored.append((sc, path, size))
    scored.sort(key=lambda t: (-t[0], t[1]))
    picked: list[dict[str, Any]] = []
    seen: set[str] = set()
    for sc, path, size in scored:
        if path in seen:
            continue
        seen.add(path)
        picked.append({"path": path, "size": size, "score": sc})
        if len(picked) >= limit:
            break
    return picked

File: frontend/js/project-ingest.js
/**
 * Regras compartilhadas de ingest de repositório (Pasta local).
 * Espelho Python: backend/integrations/project_ingest.py
 */

export const PROJECT_MAP_NAME = "__project_map.txt";
export const MAX_CONTENT_FILES = 12;
export const MAX_FILE_BYTES = 200 * 1024;
export const MAX_CONTENT_CHARS = 200000;
export const MAX_MAP_LINES = 4000;
/** 1 mapa + 12 conteúdos */
export const MAX_ATTACHMENTS = MAX_CONTENT_FILES + 1;

export const IGNORE_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  "vendor",
  "__pycache__",
  ".next",
  "coverage",
  ".venv",
  "venv",
  "target",
  "out",
  ".turbo",
  ".cache",
  ".idea",
  ".vscode",
  "eggs",
  ".eggs",
  ".tox",
  ".mypy_cache",
  ".ruff_cache",
  ".pytest_cache",
]);

const IGNORE_EXT = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".webp",
  ".ico",
  ".svg",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".mp3",
  ".mp4",
  ".wav",
  ".zip",
  ".gz",
  ".tar",
  ".7z",
  ".rar",
  ".exe",
  ".dll",
  ".so",
  ".dylib",
  ".bin",
  ".pdf",
  ".pyc",
  ".class",
  ".o",
  ".a",
  ".wasm",
  ".map",
]);

const HIGH_NAMES = new Set([
  "package.json",
  "package-lock.json",
  "pnpm-lock.yaml",
  "yarn.lock",
  "requirements.txt",
  "requirements-lock.txt",
  "pyproject.toml",
  "poetry.lock",
  "cargo.toml",
  "go.mod",
  "go.sum",
  "pom.xml",
  "composer.json",
  "composer.lock",
  "gemfile",
  "dockerfile",
  "docker-compose.yml",
  "docker-compose.yaml",
  "vercel.json",
  "next.config.js",
  "next.config.mjs",
  "next.config.ts",
  "nuxt.config.ts",
  "nuxt.config.js",
  "vite.config.ts",
  "vite.config.js",
  "tsconfig.json",
  "nginx.conf",
  ".env.example",
  ".env.sample",
  "openapi.yaml",
  "openapi.json",
  "swagger.yaml",
  "swagger.json",
]);

const SOURCE_EXT = new Set([
  ".py",
  ".js",
  ".ts",
  ".tsx",
  ".jsx",
  ".go",
  ".rs",
  ".java",
  ".kt",
  ".rb",
  ".php",
  ".cs",
  ".swift",
  ".c",
  ".cpp",
  ".h",
  ".hpp",
  ".vue",
  ".svelte",
  ".sql",
  ".sh",
  ".ps1",
  ".yml",
  ".yaml",
  ".toml",
  ".ini",
  ".cfg",
  ".conf",
  ".json",
  ".md",
  ".html",
  ".css",
  ".scss",
]);

/**
 * @param {string} path
 */
export function normalizePath(path) {
  return String(path || "")
    .replace(/\\/g, "/")
    .replace(/^\.\//, "")
    .replace(/^\/+/, "");
}

/**
 * @param {string} path
 */
export function pathIgnored(path) {
  const parts = normalizePath(path).split("/").filter(Boolean);
  for (const p of parts.slice(0, -1)) {
    if (IGNORE_DIRS.has(p)) return true;
  }
  const base = parts[parts.length - 1] || "";
  const lower = base.toLowerCase();
  const dot = lower.lastIndexOf(".");
  if (dot >= 0 && IGNORE_EXT.has(lower.slice(dot))) return true;
  if (lower.endsWith(".min.js") || lower.endsWith(".min.css")) return true;
  if (lower === "package-lock.json" && parts.length > 2) return true;
  return false;
}

/**
 * @param {string} path
 * @param {number} [size]
 */
export function scorePath(path, size = 0) {
  const norm = normalizePath(path);
  if (!norm || pathIgnored(norm)) return -1;
  if (size > MAX_FILE_BYTES) return -1;

  const parts = norm.split("/");
  const base = (parts[parts.length - 1] || "").toLowerCase();
  let score = 10;

  if (HIGH_NAMES.has(base)) score += 100;
  if (base.startsWith("dockerfile")) score += 90;
  if (base.startsWith("docker-compose")) score += 90;
  if (base.endsWith(".env.example") || base.endsWith(".env.sample")) score += 85;
  if (base.includes("nginx")) score += 70;

  const joined = norm.toLowerCase();
  if (/(^|\/)(routes|api|auth|middleware|controllers|handlers)(\/|$)/.test(joined)) {
    score += 60;
  }
  if (/(^|\/)(src|app|backend|frontend|server|web)(\/|$)/.test(joined)) {
    score += 25;
  }
  if (parts.length <= 2) score += 20;

  const dot = base.lastIndexOf(".");
  const ext = dot >= 0 ? base.slice(dot) : "";
  if (SOURCE_EXT.has(ext)) score += 15;

  if (/(^|\/)(test|tests|__tests__|spec|docs|doc|examples|fixtures)(\/|$)/.test(joined)) {
    score -= 40;
  }
  if (base.endsWith(".test.js") || base.endsWith(".spec.ts") || base.endsWith("_test.py")) {
    score -= 35;
  }
  if (base.endsWith(".lock") || base === "yarn.lock" || base === "pnpm-lock.yaml") {
    score -= 50;
  }

  return score;
}

/**
 * @param {{ path: string, size?: number }[]} entries
 */
export function buildProjectMap(entries, { maxLines = MAX_MAP_LINES } = {}) {
  const kept = [];
  let ignored = 0;
  for (const e of entries) {
    const path = normalizePath(e.path);
    if (!path || pathIgnored(path)) {
      ignored += 1;
      continue;
    }
    kept.push({ path, size: Number(e.size) || 0 });
  }
  kept.sort((a, b) => a.path.localeCompare(b.path));
  const shown = kept.slice(0, maxLines);
  const omitted = kept.length - shown.length;
  const lines = [
    `# Mapa do repositório (${kept.length} arquivos após filtros; ${ignored} ignorados)`,
    `# Amostra de conteúdo anexada à parte (arquivos prioritários para pentest).`,
    "",
  ];
  for (const row of shown) {
    lines.push(`${row.path}\t${row.size}`);
  }
  if (omitted > 0) {
    lines.push(`… (+${omitted} omitidos do mapa)`);
  }
  return {
    text: lines.join("\n"),
    keptCount: kept.length,
    ignoredCount: ignored,
    totalSeen: entries.length,
  };
}

/**
 * @param {{ path: string, size?: number }[]} entries
 */
export function pickContentPaths(entries, { limit = MAX_CONTENT_FILES } = {}) {
  const scored = [];
  for (const e of entries) {
    const path = normalizePath(e.path);
    const size = Number(e.size) || 0;
    const score = scorePath(path, size);
    if (score < 0) continue;
    scored.push({ path, size, score });
  }
  scored.sort((a, b) => b.score - a.score || a.path.localeCompare(b.path));
  const picked = [];
  const seen = new Set();
  for (const row of scored) {
    if (seen.has(row.path)) continue;
    seen.add(row.path);
    picked.push(row);
    if (picked.length >= limit) break;
  }
  return picked;
}

/**
 * Ingest de FileList do browser (Pasta do projeto).
 * Inventário completo no mapa + leitura só dos arquivos prioritários.
 * @param {FileList|File[]|null} fileList
 * @param {{
 *   addAttachment: (name: string, content: string) => void,
 *   clearAttachments?: () => void,
 *   toast?: (msg: string, type?: string) => void,
 *   onProgress?: (label: string) => void,
 * }} hooks
 */
export async function ingestProjectFiles(fileList, hooks) {
  const list = fileList || [];
  const n = list.length || 0;
  if (!n) {
    hooks.toast?.("Nenhuma pasta selecionada", "error");
    return { map: null, attached: 0, picks: [], folderName: "", files: [] };
  }

  const firstRel = normalizePath(list[0]?.webkitRelativePath || list[0]?.name || "");
  const folderName = firstRel.split("/")[0] || "pasta";

  hooks.onProgress?.(`Indexando “${folderName}” (${n} itens)…`);
  hooks.toast?.(`Indexando pasta (${n} itens)…`, "info");
  hooks.clearAttachments?.();

  /** @type {Map<string, File>} */
  const byPath = new Map();
  for (let i = 0; i < n; i++) {
    const file = list[i];
    if (!file) continue;
    const rel = normalizePath(file.webkitRelativePath || file.name);
    if (!rel) continue;
    byPath.set(rel, file);
    if (i > 0 && i % 400 === 0) {
      hooks.onProgress?.(`Indexando “${folderName}”… ${i}/${n}`);
      await new Promise((r) => setTimeout(r, 0));
    }
  }

  const entries = [];
  for (const [path, file] of byPath) {
    entries.push({ path, size: file.size || 0 });
  }

  hooks.onProgress?.(`Montando mapa de “${folderName}”…`);
  const map = buildProjectMap(entries);
  hooks.addAttachment(PROJECT_MAP_NAME, map.text.slice(0, MAX_CONTENT_CHARS));

  const picks = pickContentPaths(entries);
  const files = [];
  let attached = 0;
  for (const pick of picks) {
    const file = byPath.get(pick.path);
    if (!file) continue;
    if (file.size > MAX_FILE_BYTES) continue;
    hooks.onProgress?.(`Lendo ${pick.path}…`);
    const text = await file.text().catch(() => "");
    if (!text) continue;
    hooks.addAttachment(pick.path, text.slice(0, MAX_CONTENT_CHARS));
    files.push(pick.path);
    attached += 1;
  }

  const msg =
    `Pasta anexada: ${map.keptCount} no mapa · ${attached} arquivos lidos` +
    (map.ignoredCount ? ` · ${map.ignoredCount} ignorados` : "") +
    ` (de ${map.totalSeen})`;
  hooks.toast?.(msg, "info");
  return {
    map,
    attached,
    picks,
    folderName,
    files,
    totalSeen: map.totalSeen,
    keptCount: map.keptCount,
    ignoredCount: map.ignoredCount,
  };
}

function readAllDirectoryEntries(reader) {
  return new Promise((resolve, reject) => {
    const all = [];
    const pump = () => {
      reader.readEntries(
        (batch) => {
          if (!batch.length) {
            resolve(all);
            return;
          }
          all.push(...batch);
          pump();
        },
        reject
      );
    };
    pump();
  });
}

function entryToFile(entry, relativePath) {
  return new Promise((resolve, reject) => {
    entry.file(
      (file) => {
        try {
          Object.defineProperty(file, "webkitRelativePath", {
            value: relativePath,
            configurable: true,
          });
        } catch {
          /* ignore */
        }
        resolve(file);
      },
      reject
    );
  });
}

/**
 * Percorre FileSystemEntry (pasta/arquivo) arrastado do SO — estilo Cursor.
 * @param {FileSystemEntry} entry
 * @param {string} parentPath
 * @param {File[]} out
 */
async function walkDroppedEntry(entry, parentPath, out) {
  if (!entry) return;
  if (entry.isFile) {
    const rel = parentPath ? `${parentPath}/${entry.name}` : entry.name;
    const file = await entryToFile(entry, rel);
    out.push(file);
    if (out.length % 400 === 0) {
      await new Promise((r) => setTimeout(r, 0));
    }
    return;
  }
  if (!entry.isDirectory) return;
  const nextPrefix = parentPath ? `${parentPath}/${entry.name}` : entry.name;
  const reader = entry.createReader();
  const children = await readAllDirectoryEntries(reader);
  for (const child of children) {
    await walkDroppedEntry(child, nextPrefix, out);
  }
}

/**
 * Coleta arquivos/pastas de um DataTransfer (drag-and-drop).
 * Preferência: webkitGetAsEntry (pastas recursivas); fallback: files.
 * @param {DataTransfer|null|undefined} dt
 * @returns {Promise<File[]>}
 */
export async function collectDroppedProjectFiles(dt) {
  if (!dt) return [];
  const out = [];
  const items = dt.items ? [...dt.items] : [];

  if (items.length) {
    const entries = [];
    for (const item of items) {
      if (item.kind !== "file") continue;
      const entry = typeof item.webkitGetAsEntry === "function" ? item.webkitGetAsEntry() : null;
      if (entry) entries.push(entry);
      else {
        const f = item.getAsFile?.();
        if (f) out.push(f);
      }
    }
    for (const entry of entries) {
      await walkDroppedEntry(entry, "", out);
    }
    if (out.length) return out;
  }

  const files = dt.files ? [...dt.files] : [];
  return files;
}
